File: src/log-file-scanner.js
/**
 * Log File Scanner Module
 * Handles scanning log directories and reading log files
 * Responsibility: File system operations ONLY - NO DATABASE CODE
 * Uses service object pattern to encapsulate file scanning operations
 */

const fs = require('fs');
const path = require('path');
const readline = require('readline');

/**
 * Log File Scanner Service
 * Encapsulates all file scanning and parsing operations
 */
class LogFileScannerService {
  constructor(parserService, logger = console.log) {
    this.parserService = parserService;
    this.logger = logger;
  }
  
  /**
   * Find all NEUF log files in a directory
   * Scans only the top level (non-recursive)
   */
  findNeufLogFiles(directory) {
    const files = [];
    
    const scan = (dir) => {
      try {
        const entries = fs.readdirSync(dir, { withFileTypes: true });
        
        for (const entry of entries) {
          const fullPath = path.join(dir, entry.name);
          
          if (entry.isFile()) {
            if (/^(NEUF|neuf)-.*\.log/i.test(entry.name)) {
              files.push(fullPath);
            }
          }
        }
      } catch (error) {
        this.logger(`  Warning: Cannot read directory ${dir}: ${error.message}`);
      }
    };
    
    scan(directory);
    return files.sort();
  }
  
  /**
   * Scan all log files, parse each line directly to structured log objects,
   * and deliver them in batches via callback. No in-memory accumulation.
   * Sorting is delegated to SQLite when inserting from temp table into logs.
   * @param {string} logFolderPath - Path to log folder
   * @param {Function} onBatchReady - Callback receiving each batch of parsed log objects
   * @param {string[]|null} precomputedFiles - Optional pre-computed file list (avoids double scan)
   * @returns {Promise<number>} Total number of parsed log entries
   */
  async parseFiles(logFolderPath, onBatchReady, precomputedFiles = null) {
    this.logger('🔍 Scanning log folder:', logFolderPath);

    const logFiles = precomputedFiles || this.findNeufLogFiles(logFolderPath);

    if (logFiles.length === 0) {
      this.logger('❌ No NEUF-*.log.* files found.');
      return 0;
    }

    this.logger(`📁 Found ${logFiles.length} log file(s):`);
    logFiles.forEach(file => this.logger(`  - ${path.relative(logFolderPath, file)}`));
    this.logger('');

    const batchSize = 1000;
    let batch = [];
    let totalLines = 0;
    let totalEntries = 0;

    for (const file of logFiles) {
      const filename = path.basename(file);
      this.logger(`📄 Processing: ${path.relative(logFolderPath, file)}...`);

      const fileStream = fs.createReadStream(file, { encoding: 'utf8' });
      const rl = readline.createInterface({
        input: fileStream,
        crlfDelay: Infinity
      });

      let currentEntry = null;
      let fileLines = 0;
      let fileEntries = 0;

      for await (const line of rl) {
        totalLines++;
        fileLines++;

        const lineStart = this.parserService.detectLogLineStart(line);

        if (lineStart) {
          // Flush previous entry
          if (currentEntry) {
            const logObj = this.parserService.parseLine(
              currentEntry.timestamp + ' ' + currentEntry.rawContent,
              currentEntry.filename
            );
            if (logObj) {
              batch.push(logObj);
              totalEntries++;
              fileEntries++;

              if (batch.length >= batchSize) {
                onBatchReady(batch);
                batch = [];
              }
            }
          }

          const rawContent = line.substring(lineStart.timestamp.length).trim();
          currentEntry = { filename, timestamp: lineStart.timestamp, rawContent };
        } else {
          // Continuation line - append to current entry's rawContent
          if (currentEntry) {
            currentEntry.rawContent += '\n' + line;
          }
        }
      }

      // Flush last entry for this file
      if (currentEntry) {
        const logObj = this.parserService.parseLine(
          currentEntry.timestamp + ' ' + currentEntry.rawContent,
          currentEntry.filename
        );
        if (logObj) {
          batch.push(logObj);
          totalEntries++;
          fileEntries++;
        }
      }

      this.logger(`  ✅ ${fileLines} lines / ${fileEntries} entries.`);
    }

    // Flush remaining batch
    if (batch.length > 0) {
      onBatchReady(batch);
    }

    this.logger('');
    this.logger(`✅ Scan complete! Total lines: ${totalLines.toLocaleString()}, entries: ${totalEntries.toLocaleString()}`);
    this.logger('');

    return totalEntries;
  }

}
// Note: parseAndIndexLogFiles was removed -- it was unused dead code (no callers in the repository).

module.exports = {
  LogFileScannerService
};

File: python-port/src/log_file_scanner.py
"""
Log File Scanner Module (Python port of src/log-file-scanner.js)
Handles scanning log directories and reading log files.
Responsibility: File system operations ONLY - NO DATABASE CODE.
"""

import os
import re


class LogFileScannerService:
    """
    Log File Scanner Service.
    Encapsulates all file scanning and parsing operations.
    Port of JavaScript LogFileScannerService in src/log-file-scanner.js.
    """

    _NEUF_FILE_RE = re.compile(r'^neuf-.*\.log', re.IGNORECASE)

    def __init__(self, parser_service, logger=print):
        """
        @param parser_service: LogParserService instance
        @param logger: Callable logger
        """
        self.parser_service = parser_service
        self.logger = logger

    def find_neuf_log_files(self, directory):
        """
        Find all NEUF log files in a directory (top level only, non-recursive).
        Files must match pattern: NEUF-*.log (case-insensitive).
        Returns sorted list of full absolute file paths.

        @param directory: Path to directory to scan
        @returns: Sorted list of absolute file paths
        """
        files = []
        try:
            for entry in os.scandir(directory):
                if entry.is_file() and self._NEUF_FILE_RE.match(entry.name):
                    files.append(entry.path)
        except OSError as e:
            self.logger(f'  Warning: Cannot read directory {directory}: {e}')
        return sorted(files)

    async def parse_files(self, log_folder_path, on_batch_ready, precomputed_files=None):
        """
        Scan all log files, parse each line, deliver batches via callback.

        @param log_folder_path: Path to log folder
        @param on_batch_ready: Callable(batch: list[dict])
        @param precomputed_files: Optional pre-computed file list
        @returns: Dict { 'totalEntries': int, 'fileStats': [...] }
        """
        self.logger(f'🔍 Scanning log folder: {log_folder_path}')

        log_files = precomputed_files if precomputed_files is not None else \
                    self.find_neuf_log_files(log_folder_path)

        if not log_files:
            self.logger('❌ No NEUF-*.log files found.')
            return {'totalEntries': 0, 'fileStats': []}

        self.logger(f'📁 Found {len(log_files)} log file(s):')
        for f in log_files:
            self.logger(f'  - {os.path.relpath(f, log_folder_path)}')
        self.logger('')

        batch_size    = 1000
        batch         = []
        total_lines   = 0
        total_entries = 0
        file_stats    = []

        for filepath in log_files:
            filename    = os.path.basename(filepath)
            file_lines  = 0
            file_entries = 0
            current_entry = None

            self.logger(f'📄 Processing: {os.path.relpath(filepath, log_folder_path)}...')

            with open(filepath, encoding='utf-8', errors='replace') as fh:
                for line in fh:
                    line = line.rstrip('\n').rstrip('\r')
                    total_lines += 1
                    file_lines  += 1

                    line_start = self.parser_service.detect_log_line_start(line)

                    if line_start:
                        # Flush previous entry
                        if current_entry is not None:
                            log_obj = self.parser_service.parse_line(
                                current_entry['timestamp'] + ' ' + current_entry['raw'],
                                current_entry['filename']
                            )
                            if log_obj is not None:
                                batch.append(log_obj)
                                total_entries  += 1
                                file_entries   += 1

                                if len(batch) >= batch_size:
                                    on_batch_ready(batch)
                                    batch = []

                        raw_content = line[len(line_start['timestamp']):].strip()
                        current_entry = {
                            'filename': filename,
                            'timestamp': line_start['timestamp'],
                            'raw': raw_content,
                        }
                    else:
                        if current_entry is not None:
                            current_entry['raw'] += '\n' + line

            # Flush last entry for this file
            if current_entry is not None:
                log_obj = self.parser_service.parse_line(
                    current_entry['timestamp'] + ' ' + current_entry['raw'],
                    current_entry['filename']
                )
                if log_obj is not None:
                    batch.append(log_obj)
                    total_entries  += 1
                    file_entries   += 1

            self.logger(f'  ✅ {file_lines} lines / {file_entries} entries.')
            file_stats.append({
                'filename': os.path.relpath(filepath, log_folder_path),
                'lines':    file_lines,
                'entries':  file_entries,
            })

        # Flush remaining batch
        if batch:
            on_batch_ready(batch)

        self.logger('')
        self.logger(
            f'✅ Scan complete! Total lines: {total_lines:,}, entries: {total_entries:,}'
        )
        self.logger('')

        return {'totalEntries': total_entries, 'fileStats': file_stats}
